clients/da-vinci-client/src/main/java/com/linkedin/davinci/validation/KafkaDataIntegrityValidator.java

```diff
@@ -62,12 +119,41 @@ public void validateMessage(
 
   /**
    * For a given partition, find all the producers that has written to this partition and update the offsetRecord using
-   * segment information.
-   * @param partition
-   * @param offsetRecord
+   * segment information. Prior to this, the state which is expired according to {@link #maxAgeInMs} will be cleared.
+   *
+   * @param partition to extract info for
+   * @param offsetRecord to modify
    */
   public void updateOffsetRecordForPartition(int partition, OffsetRecord offsetRecord) {
-    producerTrackerMap.values().forEach(producerTracker -> producerTracker.updateOffsetRecord(partition, offsetRecord));
+    if (this.maxAgeInMs != DISABLED) {
+      clearExpiredStateAndUpdateOffsetRecordForPartition(partition, offsetRecord);
+    } else {
+      Iterator<ProducerTracker> iterator = this.producerTrackerMap.values().iterator();
+      while (iterator.hasNext()) {
+        iterator.next().updateOffsetRecord(partition, offsetRecord);
+      }
+    }
+  }
+
+  void clearExpiredStateAndUpdateOffsetRecordForPartition(int partition, OffsetRecord offsetRecord) {
+    long minimumRequiredRecordProducerTimestamp = offsetRecord.getMaxMessageTimeInMs() - this.maxAgeInMs;
+    int numberOfClearedGUIDs = 0;
+    Iterator<ProducerTracker> iterator = this.producerTrackerMap.values().iterator();
+    ProducerTracker producerTracker;
+    while (iterator.hasNext()) {
+      producerTracker = iterator.next();
+      if (producerTracker.clearExpiredState(partition, minimumRequiredRecordProducerTimestamp)
+          && producerTracker.isEmpty()) {
+        iterator.remove();
+        offsetRecord.setProducerPartitionState(producerTracker.getProducerGUID(), null);
```

Comment:
if `producerTracker.clearExpiredState(partition, minimumRequiredRecordProducerTimestamp)` returns true, that means this specific guid is stale for this specific partition, we can clean up the corresponding `OffsetRecord`, which is per partition, right?
Also when cleaning up the OffsetRecord, shall we remove that producer guid completely instead of setting it to be `null`?

```
public java.util.Map<java.lang.CharSequence,com.linkedin.venice.kafka.protocol.state.ProducerPartitionState> producerStates;
```
doesn't support `null` value, so I suspect `null` could break the serialization.

Reply:
I think you're right... putting null here seems messy. I'll change it to removing from the map, which seems cleaner IMHO.